test-app/tests/integration/components/headless-form-validation-test.gts

```diff
@@ -0,0 +1,833 @@
+                  },
+                ]
+              : undefined;
+
+          await render(<template>
+            <HeadlessForm
+              @data={{data}}
+              @validate={{formValidateCallback}}
+              as |form|
+            >
+              <form.field
+                @name="firstName"
+                @validate={{fieldValidateCallback}}
+                as |field|
+              >
+                <field.label>First Name</field.label>
+                <field.input data-test-first-name />
+                <field.errors data-test-first-name-errors as |errors|>
+                  {{#each errors as |e index|}}
+                    <div data-test-error={{index}}>
+                      <div data-test-error-type>
+                        {{e.type}}
+                      </div>
+                      <div data-test-error-value>
+                        {{e.value}}
+                      </div>
+                      <div data-test-error-message>
+                        {{e.message}}
+                      </div>
+                    </div>
+                  {{/each}}
+                </field.errors>
+              </form.field>
+              <form.field @name="lastName" as |field|>
+                <field.label>Last Name</field.label>
+                <field.input data-test-last-name />
+                <field.errors data-test-last-name-errors />
+              </form.field>
+              <button type="submit" data-test-submit>Submit</button>
+            </HeadlessForm>
+          </template>);
+
+          await click('[data-test-submit]');
+
+          assert
+            .dom('[data-test-first-name-errors] [data-test-error]')
+            .exists({ count: 2 });
+
+          assert
+            .dom(
+              '[data-test-first-name-errors] [data-test-error="0"] [data-test-error-type]'
+            )
+            .hasText('uppercase');
+          assert
+            .dom(
+              '[data-test-first-name-errors] [data-test-error="0"] [data-test-error-value]'
+            )
+            .hasText('foo');
+          assert
+            .dom(
+              '[data-test-first-name-errors] [data-test-error="0"] [data-test-error-message]'
+            )
+            .hasText('First name must be upper case!');
+
+          assert
+            .dom(
+              '[data-test-first-name-errors] [data-test-error="1"] [data-test-error-type]'
+            )
+            .hasText('notFoo');
+          assert
+            .dom(
+              '[data-test-first-name-errors] [data-test-error="1"] [data-test-error-value]'
+            )
+            .hasText('foo');
+          assert
+            .dom(
+              '[data-test-first-name-errors] [data-test-error="1"] [data-test-error-message]'
+            )
+            .hasText('Foo is an invalid first name!');
+
+          assert.dom('[data-test-last-name-errors]').doesNotExist();
+        });
+      })
+    );
+  });
+});
```

Comment:
Maybe this is a bit redundant or not necessary, but I'm curious what you think.

Would it be helpful to include a test that _does not_ use the field components directly, but provides their own input styled / opinionated components?  This is mostly to cover my exact use case, where we will use controls from a different addon rather than style the controls provided by headless-form.

```js
await render(
  <template>
    <HeadlessForm @data={{data}} as |form|>
      <form.field
        @name="firstName"
        @validate={{validateCallback}}
        as |field|
      >
        <!-- Verify validation works as expected for this -->
        <CustomInputComponent
          @label="First Name"
          id={{field.id}}
          value={{field.value}}
        />
        <field.errors data-test-first-name-errors />
      </form.field>
      <button type="submit" data-test-submit>Submit</button>
    </HeadlessForm>
  </template>
);
```

```hbs
<!-- CustomInputComponent -->
<!-- not complete example -->
<div>
  <label>{{@label}}</label>
  <input ...attributes />
</div>
```

Having an example of using custom control elements may help when it comes adoption time as well (i.e., have a test/docs to look at how to do this exact case).

Reply:
I too would like this...but maybe this can be a follow up since this PR is getting quite large already.

Reply:
Yeah, we should definitely add something like that into our future docs. For the test here, the input (whether that's built-in or custom) does not really play a role with validation, so the test wouldn't have much value here beyond documentation purposes I think.

However I am just realizing that using `field.errors` is actually required to use, otherwise you won't get access to the validation errors. Previously `field.errors` was just an array of errors, but now (in one of the later changes) I refactored this yield a component that actually renders a div, and that one yields the raw errors (see [example](https://github.com/CrowdStrike/ember-headless-form/pull/21/files#diff-aa61a3f8763adc87f435e3907b140da6f0c4f85fb8b1747114c4163d4bb64a0aR355-R380)). The reason was yesterday's a11y session basically, as this allows us to bake some of the a11y best practices into that div (using `aria-errormessage` to refer to is), see [this test](https://github.com/CrowdStrike/ember-headless-form/pull/21/files#diff-aa61a3f8763adc87f435e3907b140da6f0c4f85fb8b1747114c4163d4bb64a0aR270-R321). So if this were truly headless, then you wouldn't have anything to add the a11y markup to.

So given that, I wonder if there might be a need to not even use that, but render 100% of the markup yourself. In which case we would need to expose the raw errors directly again, maybe as `field.validationErrors`? Or rename what we have now from `field.errors` to something else? :think:

I think we have to block this PR though, we will still have a good amount of iterations before even thinking about using this and cutting a release...

Reply:
as long as https://developer.mozilla.org/en-US/docs/Web/API/HTMLObjectElement/validity is supported, in the future, we're 💯